File: aoc/2024/aoc202124.hs
module AOC202124 where

import Data.List(sort)

import AOC

aoc = AOC {
    day="../../2021/input/24",
    aocTests=[],
    aocCode=Code {
        codeParse=parse,
        codeParse2=parse,
        codeTest=result,
        codeTest2=result2,
        codeResult=result,
        codeResult2=result2
        }
    }

{-
inp w
mul x 0
add x z
mod x 26
div z A  -- A is 1 or 26
add x B
eql x w
eql x 0
mul y 0
add y 25
mul y x
add y 1
mul z y
mul y 0
add y w
add y C
mul y x
add z y

if input = (z mod 26) + B then
  z = z div A
else
  z = (z div A)*26 + C + input
end if

If A = 1, the if condition is always false, since B >= 10 whenever A = 1,
and input_1+C is pushed onto z.

If A = 26, the if condition must be made true.
input_1+C is popped off of z, and the condition is input_1+C = input-B,
determining input_1 and input.

-}

parse = zip [1..] . p . words
  where
    p [] = []
    p ("inp":"w":"mul":"x":"0":"add":"x":"z":"mod":"x":"26":"div":"z":a
       :"add":"x":b
       :"eql":"x":"w":"eql":"x":"0":"mul":"y":"0":"add":"y":"25":"mul":"y":"x"
       :"add":"y":"1":"mul":"z":"y":"mul":"y":"0":"add":"y":"w":"add":"y":c
       :"mul":"y":"x":"add":"z":"y":rest) =
            (read a :: Int,read b :: Int,read c :: Int) : p rest


fromDigits :: [(Int,Int)] -> Int
fromDigits = foldl collect 0 . sort
  where collect a (_,b) = a*10+b

find :: (Int -> (Int,Int)) -> [(Int,(Int,Int,Int))] -> Int
find chooseInputs = fromDigits . snd . foldl collect ([],[])
  where
    collect (z,done) (digit,(1,b,c)) = ((digit,c):z,done)
    collect ((digit1,c):z,done) (digit,(26,b,_)) =
        (z,(digit,input):(digit1,input1):done)
      where (input1,input) = chooseInputs (b+c)

result = find chooseInputs
  where chooseInputs bc | bc >= 0 = (9-bc,9)
                        | otherwise = (9,9+bc)

result2 = find chooseInputs
  where chooseInputs bc | bc >= 0 = (1,1+bc)
                        | otherwise = (1-bc,1)

File: aoc/2025/aoc202124.go
package main

func init() {
	Register(&aoc202124{
		AOC: AOC{
			Day:           24,
			InputFilename: "../2021/input/24.txt",
			Tests: []Test{
				Test{
					Input: `
`,
					Part1: "",
					Part2: "",
				},
			},
		},
	})
}

type aoc202124 struct {
	AOC
}

func (aoc *aoc202124) alu(input *Input) Seq[[3]int] {
	expect := func(s string) {
		if word, ok := input.Word(); !ok || word != s {
			panic("bad input")
		}
	}
	return func(yield func([3]int) bool) {
		for range 14 {
			var alu [3]int
			expect("inp") // w = IN
			expect("w")
			expect("mul") // x = 0        -  x = top
			expect("x")
			expect("0")
			expect("add") // x = z
			expect("x")
			expect("z")
			expect("mod") // x = z mod 26
			expect("x")
			expect("26")
			expect("div") // z = z div A - if A = 26, pop
			expect("z")
			alu[0], _ = input.Int()
			expect("add") // x = x + B   - x = top + B
			expect("x")
			alu[1], _ = input.Int()
			expect("eql") // x = w eq x  - x = IN = top + B,
			expect("x")
			expect("w")
			expect("eql") // x = x eq 0  - x = IN /= top + B,
			expect("x")
			expect("0")
			expect("mul") // y = 0
			expect("y")
			expect("0")
			expect("add") // y = 25
			expect("y")
			expect("25")
			expect("mul") // y = 25*x
			expect("y")
			expect("x")
			expect("add") // y = 25*x+1
			expect("y")
			expect("1")
			expect("mul") // z = z*y     - if IN /= top + B,
			//                           -   always true when A = 1
			//                           -   must be made false
			//                           -   when A = 26
			//                           - then push IN+C
			expect("z")
			expect("y")
			expect("mul") // y = 0
			expect("y")
			expect("0")
			expect("add") // y = w
			expect("y")
			expect("w")
			expect("add") // y = y + C
			expect("y")
			alu[2], _ = input.Int()
			expect("mul") // y = y*x
			expect("y")
			expect("x")
			expect("add") // z = z + y
			expect("z")
			expect("y")
			if !yield(alu) {
				return
			}
		}
	}
}

func (aoc *aoc202124) result(part1 bool, input *Input) string {
	z := make([][2]int, 0, 14)
	digit := 0
	result := [14]byte{}
	for alu := range aoc.alu(input) {
		if alu[0] == 1 {
			z = append(z, [2]int{digit, alu[2]})
		} else {
			x := z[len(z)-1]
			z = z[:len(z)-1]
			b := alu[1]
			c := x[1]
			// result[digit] - b = result[x[0]] + c
			if part1 {
				// maximize result[x[0]]
				for r := 9; r >= 1; r-- {
					if r+b+c < 10 {
						result[x[0]] = '0' + byte(r)
						result[digit] = '0' + byte(r+b+c)
						break
					}
				}
			} else {
				// minimize result[x[0]]
				for r := 1; r < 10; r++ {
					if r+b+c > 0 {
						result[x[0]] = '0' + byte(r)
						result[digit] = '0' + byte(r+b+c)
						break
					}
				}
			}
		}
		digit++
	}
	return string(result[:])
}

func (aoc *aoc202124) Part1(input *Input) string {
	return aoc.result(true, input)
}

func (aoc *aoc202124) Part2(input *Input) string {
	return aoc.result(false, input)
}
